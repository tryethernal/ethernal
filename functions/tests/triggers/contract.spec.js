jest.mock('axios', () => ({
    get: jest.fn()
}));
const axios = require('axios');
const { ethers } = require('../mocks/ethers');
const index = require('../../index');
const Helper = require('../helper');
const AmalfiContract = require('../fixtures/AmalfiContract.json');
const ERC20_ABI = require('../fixtures/ERC20_ABI.json');
const Transaction = {"blockHash":"0x50b10286540941286570588a16a59cc73fac0f2d8213ad49e6614d1985fd6d82","data":"0xba118f6300000000000000000000000063606c22157476da3b26ad1c2eae573d0387d7330000000000000000000000000000000000000000000000000000000065a55d7d","accessList":[],"transactionIndex":0,"confirmations":1,"type":2,"nonce":2,"gasLimit":"6721975","r":"0x22ab3ac486de70bfe6735bae92a38e18a37bbfbe09a5facc1c8bb48967a802ee","s":"0x6ad0bebea8f80ff739a6cf9f75c057292ee2e29ded602734e5d496bd063797e3","chainId":31337,"v":1,"blockNumber":14008967,"from":"0x2d481eeb2ba97955cd081cf218f453a817259ab1","receipt":{"blockHash":"0x50b10286540941286570588a16a59cc73fac0f2d8213ad49e6614d1985fd6d82","logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","transactionIndex":0,"confirmations":1,"transactionHash":"0xaa82e74b71fb20503205beaffb60a38b08c6766a66dae1477b479df0e05f1bbe","gasUsed":"46419","blockNumber":14008967,"cumulativeGasUsed":"46419","from":"0x2d481eeb2ba97955cd081cf218f453a817259ab1","to":"0x63606c22157476da3b26ad1c2eae573d0387d733","logs":[],"byzantium":true,"status":1},"to":"0x63606c22157476da3b26ad1c2eae573d0387d733","value":"0","hash":"0xaa82e74b71fb20503205beaffb60a38b08c6766a66dae1477b479df0e05f1bbe","gasPrice":"76869329841","timestamp":1642264024,"methodDetails":{"signature":"setMaturity(address payee, uint256 maturity)","name":"setMaturity","label":"setMaturity(\n\taddress payee: 0x63606C22157476Da3B26Ad1c2EAE573D0387D733,\n\tuint256 maturity: 1705336189\n)"}};
let helper;

describe('processContract', () => {
    beforeEach(() => {
        helper = new Helper(process.env.GCLOUD_PROJECT);
    });

    it('Should store token info, tag it as erc20, if the workspace is public', async () => {
        await helper.workspace
            .set({ public: true }, { merge: true });

        const data = helper.test.firestore.makeDocumentSnapshot({
            address: Transaction.to,
            abi: ERC20_ABI
        }, `users/123/workspaces/hardhat/contracts/${Transaction.to}`);

        await helper.test.wrap(index.processContract)(data, { params: { userId: '123', workspaceName: 'hardhat' }});
    
        const doc = await helper.workspace.collection('contracts').doc(Transaction.to).get();
        expect(doc.data()).toEqual({
            patterns: ['erc20'],
            abi: ERC20_ABI,
            processed: true,
            token: {
                decimals: 18,
                name: 'Ethernal',
                symbol: 'ETL'
            }
        });
    });

    it('Should store token info, tag it as erc20 & proxy, if the workspace is public', async () => {
        await helper.workspace
            .set({ public: true }, { merge: true });

        const data = helper.test.firestore.makeDocumentSnapshot({
            address: Transaction.to,
            abi: [{ my: 'function' }]
        }, `users/123/workspaces/hardhat/contracts/${Transaction.to}`);

        await helper.test.wrap(index.processContract)(data, { params: { userId: '123', workspaceName: 'hardhat' }});
    
        const doc = await helper.workspace.collection('contracts').doc(Transaction.to).get();
        expect(doc.data()).toEqual({
            patterns: ['erc20', 'proxy'],
            abi: [{ my: 'function' }],
            processed: true,
            token: {
                decimals: 18,
                name: 'Ethernal',
                symbol: 'ETL'
            }
        });
    });

    it('Should process associated existing transactions', async () => {
        await helper.workspace
            .collection('transactions')
            .doc(Transaction.hash)
            .set(Transaction);

        const data = helper.test.firestore.makeDocumentSnapshot({
            address: Transaction.to,
            abi: AmalfiContract.artifact.abi,
        }, `users/123/workspaces/hardhat/contracts/${Transaction.to}`);

        await helper.test.wrap(index.processContract)(data, { params: { userId: '123', workspaceName: 'hardhat' }});

        const transactionDoc = await helper.workspace
            .collection('transactions')
            .doc(Transaction.hash)
            .get();

        expect(transactionDoc.data()).toMatchSnapshot();
    });

    it('Should import and link proxy contract', async () => {
        await helper.workspace.collection('contracts').doc('0x125').set({
            address: '0x125',
            imported: true
        });
        const data = helper.test.firestore.makeDocumentSnapshot({
            address: '0x125',
        }, 'users/123/workspaces/hardhat/contracts/0x125');

        axios.get.mockImplementation(() => ({
            data: {
                message: 'OK',
                result: [{
                    ContractName: 'Contract',
                    Proxy: '1',
                    Implementation: '0xabcd',
                    ABI: JSON.stringify([{ my: 'function' }])
                }]
            }
        }));

        const result = await helper.test.wrap(index.processContract)(data, { params: { userId: '123', workspaceName: 'hardhat' }});

        const doc = await helper.workspace.collection('contracts').doc('0x125').get();
        expect(doc.data()).toEqual({
            address: '0x125',
            name: 'Contract',
            abi: [{ my: 'function' }],
            proxy: '0xabcd',
            imported: true
        });
    });

    it('Should match when contract is found locally', async () => {
        await helper.workspace.collection('contracts').doc('0x123').set({
            name: 'Contract',
            address: '0x123',
            hashedBytecode: '0x123abcd',
            abi: [{ my: 'function' }]
        });
        await helper.workspace.collection('contracts').doc('0x124').set({
            hashedBytecode: '0x123abcd',
            address: '0x124'
        });

        const data = helper.test.firestore.makeDocumentSnapshot({
            address: '0x124',
            hashedBytecode: '0x123abcd'
        }, 'users/123/workspaces/hardhat/contracts/0x124');

        const result = await helper.test.wrap(index.processContract)(data, { params: { userId: '123', workspaceName: 'hardhat' }});
        const doc = await helper.workspace.collection('contracts').doc('0x124').get();

        expect(doc.data()).toEqual({
            hashedBytecode: '0x123abcd',
            address: '0x124',
            name: 'Contract',
            abi: [{ my: 'function' }]
        });
    });

    it('Should match with etherscan if it is found', async () => {
        await helper.workspace.collection('contracts').doc('0x125').set({ address: '0x125', hashedBytecode: '0x123a' });
        const data = helper.test.firestore.makeDocumentSnapshot({
            address: '0x125',
            hashedBytecode: '0x123a'
        }, 'users/123/workspaces/hardhat/contracts/0x125');

        axios.get.mockImplementation(() => ({
            data: {
                message: 'OK',
                result: [{
                    ContractName: 'Contract',
                    ABI: JSON.stringify([{ my: 'function' }])
                }]
            }
        }));

        const result = await helper.test.wrap(index.processContract)(data, { params: { userId: '123', workspaceName: 'hardhat' }});

        const doc = await helper.workspace.collection('contracts').doc('0x125').get();
        expect(doc.data()).toEqual({
            address: '0x125',
            hashedBytecode: '0x123a',
            name: 'Contract',
            abi: [{ my: 'function' }]
        });
    });

    it('Should stay the same if it does not match anything', async () => {
        await helper.workspace.collection('contracts').doc('0x123').set({ address: '0x123', hashedBytecode: '0x123abcd' });
        const data = helper.test.firestore.makeDocumentSnapshot({
            address: '0x123',
            hashedBytecode: '0x123abcd'
        }, 'users/123/workspaces/hardhat/contracts/0x123');
        
        axios.get.mockImplementation(() => ({
            data: {
                message: 'NOTOK'
            }
        }));

        const result = await helper.test.wrap(index.processContract)(data, { params: { userId: '123', workspaceName: 'hardhat' }});
        const doc = await helper.workspace.collection('contracts').doc('0x123').get();
        expect(doc.data()).toEqual({
            address: '0x123',
            hashedBytecode: '0x123abcd'
        });
    });    

    it('Should not match if no hashedBytecode & ABI', async () => {
        await helper.workspace.collection('contracts').doc('0x123').set({ address: '0x123' });
        const data = helper.test.firestore.makeDocumentSnapshot({
            abi: { my: 'function' }
        }, 'users/123/workspaces/hardhat/contracts/0x123');

        const result = await helper.test.wrap(index.processContract)(data, { params: { userId: '123', workspaceName: 'hardhat' }});;

        const doc = await helper.workspace.collection('contracts').doc('0x123').get();
        expect(doc.data()).toEqual({ address: '0x123', abi: { my: 'function' } });
    });

    it('Should not match if no abi & hashedBytecode & no etherscan match', async () => {
        await helper.workspace.collection('contracts').doc('0x123').set({ address: '0x123', hashedBytecode: '0x123' });
        const data = helper.test.firestore.makeDocumentSnapshot({
            hashedBytecode: '0x123'
        }, 'users/123/workspaces/hardhat/contracts/0x123');

        axios.get.mockImplementation(() => ({
            data: {
                message: 'NOTOK'
            }
        }));

        const result = await helper.test.wrap(index.processContract)(data, { params: { userId: '123', workspaceName: 'hardhat' }});;

        const doc = await helper.workspace.collection('contracts').doc('0x123').get();
        expect(doc.data()).toEqual({ address: '0x123', hashedBytecode: '0x123' });
    });

    afterEach(() => helper.clean());
});