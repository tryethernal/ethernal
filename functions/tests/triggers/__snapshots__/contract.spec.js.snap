// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processContract Should process associated existing transactions 1`] = `
Object {
  "accessList": Array [],
  "blockHash": "0x50b10286540941286570588a16a59cc73fac0f2d8213ad49e6614d1985fd6d82",
  "blockNumber": 14008967,
  "chainId": 31337,
  "confirmations": 1,
  "data": "0xba118f6300000000000000000000000063606c22157476da3b26ad1c2eae573d0387d7330000000000000000000000000000000000000000000000000000000065a55d7d",
  "from": "0x2d481eeb2ba97955cd081cf218f453a817259ab1",
  "gasLimit": "6721975",
  "gasPrice": "76869329841",
  "hash": "0xaa82e74b71fb20503205beaffb60a38b08c6766a66dae1477b479df0e05f1bbe",
  "methodDetails": Object {
    "label": "setMaturity(
	address payee: 0x63606C22157476Da3B26Ad1c2EAE573D0387D733,
	uint256 maturity: 1705336189
)",
    "name": "setMaturity",
    "signature": "setMaturity(address payee, uint256 maturity)",
  },
  "nonce": 2,
  "r": "0x22ab3ac486de70bfe6735bae92a38e18a37bbfbe09a5facc1c8bb48967a802ee",
  "receipt": Object {
    "blockHash": "0x50b10286540941286570588a16a59cc73fac0f2d8213ad49e6614d1985fd6d82",
    "blockNumber": 14008967,
    "byzantium": true,
    "confirmations": 1,
    "cumulativeGasUsed": "46419",
    "from": "0x2d481eeb2ba97955cd081cf218f453a817259ab1",
    "gasUsed": "46419",
    "logs": Array [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "status": 1,
    "to": "0x63606c22157476da3b26ad1c2eae573d0387d733",
    "transactionHash": "0xaa82e74b71fb20503205beaffb60a38b08c6766a66dae1477b479df0e05f1bbe",
    "transactionIndex": 0,
  },
  "s": "0x6ad0bebea8f80ff739a6cf9f75c057292ee2e29ded602734e5d496bd063797e3",
  "timestamp": 1642264024,
  "to": "0x63606c22157476da3b26ad1c2eae573d0387d733",
  "tokenTransfers": Array [],
  "transactionIndex": 0,
  "type": 2,
  "v": 1,
  "value": "0",
}
`;
